### PR TITLE
issue - 1785 : Fix for the issue "FileBasedStateRepository uses the w…

### PR DESCRIPTION
**Pull request by vaibhavsrivastava07**
CustomObjectInputStream & CustomLoaderFileBasedObjectInputStream classes introduced to hanle the order of classloader for execution. The flow will first look for TCCL, if class is not loaded then it will look for Cachemanager classloader if found then it will load otherwise it will look for the latest user defined classloader. Loadmaps method is loaded lazily when requested for persistent context. This will allow us to use the classloaders provided.

### CODE REVIEW

File: ehcache-api/src/main/java/org/ehcache/spi/persistence/StateRepository.java [hunk 1]
```diff
@@ -39,7 +41,7 @@ public interface StateRepository {
    * @return a state holder
    */
   @Deprecated
-  default <K extends Serializable, V extends Serializable> StateHolder<K, V> getPersistentStateHolder(String name, Class<K> keyClass, Class<V> valueClass) {
+  default <K extends Serializable, V extends Serializable> StateHolder<K, V> getPersistentStateHolder(String name, Class<K> keyClass, Class<V> valueClass) throws CachePersistenceException {
```
Comment (AbfrmBlr) [inline ehcache-api/src/main/java/org/ehcache/spi/persistence/StateRepository.java:44]: Why we need a change in public api? Is this really required?

File: ehcache-impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java [hunk 2]
```diff
@@ -80,8 +85,12 @@ public static <T> Class<? extends Serializer<T>> asTypedSerializer() {
 
   @Override
   public void init(final StateRepository stateRepository) {
-    this.persistentState = stateRepository.getPersistentStateHolder("CompactJavaSerializer-ObjectStreamClassIndex", Integer.class, ObjectStreamClass.class, c -> true, null);
-    refreshMappingsFromStateRepository();
+    try {
+      this.persistentState = stateRepository.getPersistentStateHolder("CompactJavaSerializer-ObjectStreamClassIndex", Integer.class, ObjectStreamClass.class, c -> true, null);
+      refreshMappingsFromStateRepository();
+    }catch (CachePersistenceException e){
```
Comment (AbfrmBlr) [inline ehcache-impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java:91]: Incorrect formatting.

Comment (AbfrmBlr) [inline ehcache-impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java:91]: Not sure I understand this change here, the init throws exception and we eat it up without letting user know the init has failed. Care to elaborate?

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java [hunk 3]
```diff
@@ -100,8 +99,14 @@ public <K extends Serializable, V extends Serializable> StateHolder<K, V> getPer
                                                                                                      Class<K> keyClass,
                                                                                                      Class<V> valueClass,
                                                                                                      Predicate<Class<?>> isClassPermitted,
-                                                                                                     ClassLoader classLoader) {
-    // isClassPermitted and  classLoader are ignored because this state repository has already being read from file and cached in
+                                                                                                     ClassLoader classLoader) throws CachePersistenceException {
+    // Lazy loading of loadMaps using class loader in order TCCL, CacheManager CL, Custom Classloader
```
Comment (AbfrmBlr) [inline ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java:103]: What's the intent behind this change of moving the `loadMaps` call from constructor to this method? Why are we loading the map at every call of `getPersistentHolder`? What's the point of loading in the map, if we are calling it every time?

**Comment by AbfrmBlr** [comment]
The logic in `CustomLoaderFileBasedObjectInputStream.java` looks up for the class in user provided classloader before trying to load it from CacheManager's classloader? Am I missing something @vaibhavsrivastava07 ?

File: ehcache-api/src/main/java/org/ehcache/spi/persistence/StateRepository.java (hunk 1 above)
Comment (vaibhavsrivastava07) [reply ehcache-api/src/main/java/org/ehcache/spi/persistence/StateRepository.java:44]: We can avoid changing public API by removing CachePersistenceException and re-throwing as RuntimeException in FileBasedStateRepository:getPersistentStateHolder() following the ehcahe current coding practice.

File: ehcache-impl/src/test/java/org/ehcache/impl/serialization/CompactJavaSerializerTest.java [hunk 4]
```diff
@@ -42,8 +43,11 @@ public class CompactJavaSerializerTest {
   public void testStateHolderFailureRereadBehavior() throws ClassNotFoundException {
     StateHolder<Integer, ObjectStreamClass> stateMap = spy(new TransientStateHolder<>());
     StateRepository stateRepository = mock(StateRepository.class);
-    when(stateRepository.getPersistentStateHolder(eq("CompactJavaSerializer-ObjectStreamClassIndex"), eq(Integer.class), eq(ObjectStreamClass.class), any(), any())).thenReturn(stateMap);
-
+    try {
+      when(stateRepository.getPersistentStateHolder(eq("CompactJavaSerializer-ObjectStreamClassIndex"), eq(Integer.class), eq(ObjectStreamClass.class), any(), any())).thenReturn(stateMap);
+    }catch (CachePersistenceException e){
```
[the remark this answers is not in the document]
Comment (vaibhavsrivastava07) [reply ehcache-impl/src/test/java/org/ehcache/impl/serialization/CompactJavaSerializerTest.java:48]: The exception re-throwing as a RuntimeException in FileBasedStateRepository:getPersistentStateHolder(), this change will be reverted and formatting will be corrected.

File: ehcache-impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java (hunk 2 above)
Comment (vaibhavsrivastava07) [reply ehcache-impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java:91]: The exception is re-thrown as a RuntimeException in FileBasedStateRepository:getPersistentStateHolder(), hence this change will be reverted and formatting will be corrected.

Comment (vaibhavsrivastava07) [reply ehcache-impl/src/main/java/org/ehcache/impl/serialization/CompactJavaSerializer.java:91]: Explanation same as above comment. We will be reverting this change.

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/CustomObjectInputStream.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.ehcache.impl.persistence;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InvalidClassException;
+import java.io.ObjectStreamClass;
+import java.util.function.Predicate;
+
+public class CustomObjectInputStream extends CustomLoaderFileBasedObjectInputStream {
```
[the remark this answers is not in the document]
Comment (vaibhavsrivastava07) [reply ehcache-impl/src/main/java/org/ehcache/impl/persistence/CustomObjectInputStream.java:25]: Yes, we need this to decide the order of class loader used for class loading.
**Scenario 1 :**
    class loader passed as null --> 
        1. TCCL - It will look for TCCL
        2. Cache Manager - If not loaded from TCCL it will look into Cache Manager to load the classes
        3. latestUserDefinedLoader - If not loaded from Cache Manager it will use the Latest user defined(sun.misc.VM.latestUserDefinedLoader())

**Scenario 2 :** 
    Passing Custom class loader--> 
        1. Custom class loader - It will look for Custom class loader
        2. Cache Manager - If not loaded from Custom class loader it will look into Cache Manager to load the classes
        3. latestUserDefinedLoader - If not loaded from Cache Manager it will use the Latest user defined(sun.misc.VM.latestUserDefinedLoader())

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java (hunk 3 above)
Comment (vaibhavsrivastava07) [reply ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java:103]: To enable the Lazy loading and implement class loader ordering and passing custom class loader as parameter for de-serialization we have moved this method to getPersistentHolder(). 
getPersistentHolder() has the parameter class loader and to use this, we have moved loadMaps call from constructor to getPersistentHolder().

**Comment by vaibhavsrivastava07** [comment]
> The logic in `CustomLoaderFileBasedObjectInputStream.java` looks up for the class in user provided classloader before trying to load it from CacheManager's classloader? Am I missing something @vaibhavsrivastava07 ?



> The logic in `CustomLoaderFileBasedObjectInputStream.java` looks up for the class in user provided classloader before trying to load it from CacheManager's classloader? Am I missing something @vaibhavsrivastava07 ?
Explained in above comments for the class loading order.

**Comment by vaibhavsrivastava07** [comment]
Explained in above comments for class loading ordering.

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/CustomObjectInputStream.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (AbfrmBlr) [reply ehcache-impl/src/main/java/org/ehcache/impl/persistence/CustomObjectInputStream.java:25]: No I don't see any logic of classloading in this class, it is just doing the filtering as whether the class being loaded is permitted or not.
The class resolution is only being done in its super class, so the question remains - DO we need this kind if filtering for Filebased state repository.

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java (hunk 3 above)
Comment (AbfrmBlr) [reply ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java:103]: There is nothing "lazy" about loading the user provided classes here. All these operations are called during the init phase of `Store.Provider` initialize the serializers. Please look up the call stack of this method and also you read more about Serializers in user doc https://www.ehcache.org/documentation/3.10/serializers-copiers.html. Effectively all the user provided classes which are Keys & Values types used in cache are loaded when the Cache is initialized.

**Comment by AbfrmBlr** [comment]
Please add tests for respective change

**Comment by AbfrmBlr** [comment]
To understand this issue better, I would recommend to implement a test case in which a disk persistent Cache uses both key & value which are user defined classes. This issue wants to address the issue of classloading of user defined classes.

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java (hunk 3 above)
Comment (AbfrmBlr) [reply ehcache-impl/src/main/java/org/ehcache/impl/persistence/FileBasedStateRepository.java:103]: Ignore my previous comment, delaying loading the classes till the getPersistentHolder is the right thing to do.

File: ehcache-impl/src/main/java/org/ehcache/impl/persistence/CustomObjectInputStream.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vaibhavsrivastava07) [reply ehcache-impl/src/main/java/org/ehcache/impl/persistence/CustomObjectInputStream.java:25]: This kind predicate filtering is not required as class loading order logic is in parent class CustomLoaderFileBasedObjectInputStream. Hence We will remove CustomObjectInputStream class.

**Comment by vaibhavsrivastava07** [comment]
We will add respective test case for disk persistent Cache uses both key & value which are user defined classes.

**Comment by AbfrmBlr** [comment]
The added tests don't fail in absence of the fix, which means it does not reproduce the problem at the hand and is unrelated to the issue.

File: ehcache-impl/build.gradle [hunk 6]
```diff
@@ -50,6 +50,7 @@ task slowTest(type: Test) {
 
 dependencies {
   api project(':ehcache-core')
+  implementation project(':clustered:ehcache-common')
```
Comment (AbfrmBlr) [inline ehcache-impl/build.gradle:53]: clustered modules should not be added as a dependency in ehcache impl

Comment (AbfrmBlr) [reply ehcache-impl/build.gradle:53]: Please look into possibility of extracting the files to a "shared" module which becomes a dependency for both ehcache-impl & clustered. That would also serve the intent of reusing the existing code.

**Comment by AbfrmBlr** [comment]
As discussed in the yesterday's call , make sure to include only relevant changes in the PR.

File: clustered/ehcache-common/src/main/java/org/ehcache/clustered/common/internal/store/Util.java [hunk 7]
```diff
@@ -17,17 +17,13 @@
 package org.ehcache.clustered.common.internal.store;
 
 import org.ehcache.clustered.common.internal.util.ByteBufferInputStream;
+import org.ehcache.impl.store.FilteredObjectInputStream;
```
Comment (AbfrmBlr) [inline clustered/ehcache-common/src/main/java/org/ehcache/clustered/common/internal/store/Util.java:20]: why this change?

File: clustered/ehcache-common/build.gradle [hunk 8]
```diff
@@ -28,6 +28,7 @@ publishing.publications.withType(MavenPublication) {
 dependencies {
   api project(':ehcache-api')
   api project(':clustered:ehcache-common-api')
+  api project(':ehcache-impl')
```
Comment (AbfrmBlr) [inline clustered/ehcache-common/build.gradle:31]: Similarly, why this one?

File: clustered/ehcache-common/src/main/java/org/ehcache/clustered/common/internal/store/Util.java (hunk 7 above)
Comment (amitbharate) [reply clustered/ehcache-common/src/main/java/org/ehcache/clustered/common/internal/store/Util.java:20]: Thanks for the review. After moving FilteredObjectInputStream from clustered-common to ehcache-impl it requires a import and as currently there is no dependency of ehcache-impl in clustered common we added these 2 change. Please suggest.

**Comment by amitbharate** [comment]
> As discussed in the yesterday's call , make sure to include only relevant changes in the PR.

Thanks Abhilash. We are working on cleaning up remaining 2 unwanted commits.

File: clustered/ehcache-common/src/main/java/org/ehcache/clustered/common/internal/store/Util.java (hunk 7 above)
Comment (vaibhavsrivastava07) [reply clustered/ehcache-common/src/main/java/org/ehcache/clustered/common/internal/store/Util.java:20]: > 

Unwanted commits has been removed.

**Comment by vaibhavsrivastava07** [comment]
> As discussed in the yesterday's call , make sure to include only relevant changes in the PR.

Unwanted commits has been removed.

File: clustered/ehcache-client/build.gradle [hunk 9]
```diff
@@ -36,6 +36,7 @@ dependencies {
   implementation "org.terracotta:lease-api:$terracottaPlatformVersion"
   implementation "org.terracotta.dynamic-config.entities:dynamic-config-topology-entity-client:$terracottaPlatformVersion"
   implementation "org.terracotta:connection-api:$terracottaApisVersion"
+  implementation group: 'org.terracotta', name: 'terracotta-utilities-tools', version: parent.terracottaUtilitiesVersion
```
Comment (AbfrmBlr) [inline clustered/ehcache-client/build.gradle:39]: It will be nice to maintain consistency while adding dependencies using format 
`implemenation "org.terracotta:terracotta-utilities-tools:$terracottaUtilitiesVersion`

File: clustered/ehcache-common/build.gradle [hunk 10]
```diff
@@ -31,6 +31,7 @@ dependencies {
 
   implementation "org.terracotta:entity-common-api:$terracottaApisVersion"
   implementation "org.terracotta:runnel:$terracottaPlatformVersion"
+  implementation group: 'org.terracotta', name: 'terracotta-utilities-tools', version: parent.terracottaUtilitiesVersion
```
Comment (AbfrmBlr) [inline clustered/ehcache-common/build.gradle:34]: Similarly here as well
